aten/src/ATen/native/cuda/ScaledGroupMM.cu

```diff
@@ -41,16 +44,87 @@ C10_DIAGNOSTIC_PUSH_AND_IGNORED_IF_DEFINED("-Wunused-but-set-parameter")
 
 #include <cute/atom/mma_atom.hpp>
 #include <cutlass/gemm/dispatch_policy.hpp>
+#include <cutlass/gemm/group_array_problem_shape.hpp>
 #include <cutlass/gemm/kernel/gemm_universal.hpp>
 #include <cutlass/util/packed_stride.hpp>
 
+// Added for SM100 support
+#include <cutlass/epilogue/collective/default_epilogue.hpp>
+#include <cutlass/epilogue/thread/activation.h>
+#include <cutlass/epilogue/thread/linear_combination.h>
+#include <cutlass/gemm/kernel/tile_scheduler_params.h>
+#include <cutlass/tensor_ref.h>
+
+#if (defined(CUDA_VERSION) && CUDA_VERSION >= 12080) && \
+    (defined(CUTLASS_ARCH_MMA_SM100A_SUPPORTED) ||      \
+     defined(CUTLASS_ARCH_MMA_SM100_SUPPORTED))
+#define BUILD_SM100_KERNEL
+#endif
+
+C10_DIAGNOSTIC_POP()
 C10_DIAGNOSTIC_POP()
 C10_DIAGNOSTIC_POP()
 
+namespace at::cuda::detail {
+
+GroupCountInfo get_group_count(
+    at::Tensor mat_a,
+    at::Tensor mat_b,
+    std::optional<at::Tensor> offs) {
+  int M = mat_a.size(-2);
+  int K = mat_a.size(-1);
+  int N = mat_b.size(-1);
+  int group_count = 0;
+  GroupMMInputMatrixType type{};
+
+  if (mat_a.dim() == 2 && mat_b.dim() == 2) {
+    // if both inputs are ragged, K is dynamic, M and N come from inputs
+    group_count = offs->size(0);
+    type = GroupMMInputMatrixType::GroupMMInputMatrixType_MatrixA_2D_MatrixB_2D;
+
+    // stack on the K dimension
+    K = K / group_count;
+  } else if (mat_a.dim() == 2) {
+    group_count = mat_b.size(0);
+    type = GroupMMInputMatrixType::GroupMMInputMatrixType_MatrixA_2D_MatrixB_3D;
+    // stack on the M dimension
+    M = M / group_count;
+  } else if (mat_b.dim() == 2) {
+    group_count = mat_a.size(0);
+    type = GroupMMInputMatrixType::GroupMMInputMatrixType_MatrixA_3D_MatrixB_2D;
+    // stack on the N dimension
+    N = N / group_count;
+  } else {
+    // regular bmm
+    group_count = mat_a.size(0);
+    type = GroupMMInputMatrixType::GroupMMInputMatrixType_MatrixA_3D_MatrixB_3D;
+  }
+
+  return GroupCountInfo{M, N, K, group_count, type};
+}
+
+} // namespace at::cuda::detail
+
 namespace {
 
 using Strides = at::cuda::detail::Strides;
 
+int ceildiv(int a, int b) {
```

Comment:
we have these here: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/ceil_div.h